src/sagemaker/serve/builder/jumpstart_builder.py

```diff
@@ -234,6 +254,213 @@ def _build_for_tgi_jumpstart(self):
+        """Tune for Jumpstart Models.
+
+        Args:
+            multiple_model_copies_enabled (bool): Whether multiple model copies serving is enable by
+                this ``DLC``. Defaults to ``False``
+            max_tuning_duration (int): The maximum timeout to deploy this ``Model`` locally.
+                Default: ``1800``
+        returns:
+            Tuned Model.
+        """
+        if self.mode != Mode.LOCAL_CONTAINER:
+            logger.warning(
+                "Tuning is only a %s capability. Returning original model.", Mode.LOCAL_CONTAINER
+            )
+            return self.pysdk_model
+
+        num_shard_env_var_name = "SM_NUM_GPUS"
+        if "OPTION_TENSOR_PARALLEL_DEGREE" in self.pysdk_model.env.keys():
+            num_shard_env_var_name = "OPTION_TENSOR_PARALLEL_DEGREE"
+
+        initial_env_vars = copy.deepcopy(self.pysdk_model.env)
+        admissible_tensor_parallel_degrees = _get_admissible_tensor_parallel_degrees(
+            self.js_model_config
+        )
+        available_gpus = _get_available_gpus() if multiple_model_copies_enabled else None
+
+        benchmark_results = {}
+        best_tuned_combination = None
+        timeout = datetime.now() + timedelta(seconds=max_tuning_duration)
+        for tensor_parallel_degree in admissible_tensor_parallel_degrees:
+            if datetime.now() > timeout:
+                logger.info("Max tuning duration reached. Tuning stopped.")
+                break
+
+            sagemaker_model_server_workers = 1
+            if multiple_model_copies_enabled:
+                sagemaker_model_server_workers = int(available_gpus / tensor_parallel_degree)
+
+            self.pysdk_model.env.update(
+                {
+                    num_shard_env_var_name: str(tensor_parallel_degree),
+                    "SAGEMAKER_MODEL_SERVER_WORKERS": str(sagemaker_model_server_workers),
+                }
+            )
+
+            try:
+                predictor = self.pysdk_model.deploy(model_data_download_timeout=max_tuning_duration)
+
+                avg_latency, p90, avg_tokens_per_second = _serial_benchmark(
+                    predictor, self.schema_builder.sample_input
+                )
+                throughput_per_second, standard_deviation = _concurrent_benchmark(
+                    predictor, self.schema_builder.sample_input
+                )
+
+                tested_env = self.pysdk_model.env.copy()
+                logger.info(
+                    "Average latency: %s, throughput/s: %s for configuration: %s",
+                    avg_latency,
+                    throughput_per_second,
+                    tested_env,
+                )
+                benchmark_results[avg_latency] = [
+                    tested_env,
+                    p90,
+                    avg_tokens_per_second,
+                    throughput_per_second,
+                    standard_deviation,
+                ]
+
+                if not best_tuned_combination:
```

Comment:
Will best tuned combination ever be set? Or is this for the case of multiple model copies enabled?

Reply:
Yes, and I fixed the logic. The best tuned combination will be set